### PR TITLE
[onert] Apply ReLU6Grad to ElementwiseActivationLayer

### PR DESCRIPTION
**Pull request by zetwhite**
This PR applies ReLU6Grad to ElementwiseActivationLayer.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

issue : https://github.com/Samsung/ONE/issues/12388 
draft : https://github.com/Samsung/ONE/pull/12395

### CODE REVIEW

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 1]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
-                                        getShape(incoming), getBuffer<float>(incoming),
-                                        getShape(outgoing), getBuffer<float>(outgoing));
+          auto relu_cker = [&alpha]() {
+            if (alpha == std::numeric_limits<float>::infinity())
+              return nnfw::cker::train::ReLUGrad;
+            else if (alpha == 6.0f)
+              return nnfw::cker::train::ReLU6Grad;
+            else
+              throw std::runtime_error{"no supported relu kernel"};
+          }();
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:70]: (note) select kernel by alpha value

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 2]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
+            relu_cker(getShape(output), getBuffer<float>(output), getShape(incoming),
+                      getBuffer<float>(incoming), getShape(outgoing), getBuffer<float>(outgoing));
           };
         }
         else
         {
           throw std::runtime_error("train ElementwiseActivationLayer : This layer does not "
-                                   "suppport other ReLU except for ReLU(0-inf)");
+                                   "suppport other ReLU except for ReLU(0-inf) and ReLU6(0-6)");
         }
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:83]: ### NOTE
# WIP

After This PR, the training feature supports `ReLU6`. (not fused one) 

I also checked loss value isn't different with tensorflow. 



* TensorFlow
```
Epoch 1/5
50/50 [==============================] - 0s 617us/step - loss: 0.0794 - mean_squared_error: 0.0794
Epoch 2/5
50/50 [==============================] - 0s 542us/step - loss: 0.0586 - mean_squared_error: 0.0586
Epoch 3/5
50/50 [==============================] - 0s 533us/step - loss: 0.0505 - mean_squared_error: 0.0505
Epoch 4/5
50/50 [==============================] - 0s 542us/step - loss: 0.0461 - mean_squared_error: 0.0461
Epoch 5/5
50/50 [==============================] - 0s 513us/step - loss: 0.0431 - mean_squared_error: 0.0431
```

* onert_train
```
- learning_rate   = 0.001
- batch_size      = 20
- loss_info       = {loss = mean squared error, reduction = sum over batch size}
- optimizer       = adam
========================
Epoch 1/5 - time: 3.351ms/step - loss: [0] 0.0794
Epoch 2/5 - time: 3.375ms/step - loss: [0] 0.0586
Epoch 3/5 - time: 3.409ms/step - loss: [0] 0.0505
Epoch 4/5 - time: 3.655ms/step - loss: [0] 0.0461
Epoch 5/5 - time: 3.915ms/step - loss: [0] 0.0431
===================================
```

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 3]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
+            else
+              throw std::runtime_error{"no supported relu kernel"};
+          }();
+
+          _backward_kernel = [relu_cker](const IPortableTensor *output,
+                                         const IPortableTensor *incoming,
+                                         IPortableTensor *outgoing) {
+            relu_cker(getShape(output), getBuffer<float>(output), getShape(incoming),
+                      getBuffer<float>(incoming), getShape(outgoing), getBuffer<float>(outgoing));
           };
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:77]: (note) create `_backend_kernel`, by capturing selected `relu_cker`

**Review by ragmani — DISMISSED** [review]
LGTM

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 4]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
         }
         else
         {
           throw std::runtime_error("train ElementwiseActivationLayer : This layer does not "
-                                   "suppport other ReLU except for ReLU(0-inf)");
+                                   "suppport other ReLU except for ReLU(0-inf) and ReLU6(0-6)");
```
Comment (glistening) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:82]: (optional) It would be good not to list the supported ReLU types here. We need to update the message whenever supported list is updated. What about something like "Not supported ReLU type activation".

**Review by glistening — DISMISSED** [review]
LGTM

**Review by jyoungyun — DISMISSED** [review]
LGTM

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc (hunk 4 above)
Comment (zetwhite) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc]: Looks better!
updated :)

**Review by ragmani — APPROVED** [review]

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 5]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
         {
           cpu::ops::ElementwiseActivationLayer::configure(
             input, output, alpha, beta, cpu::ops::ElementwiseActivationType::kReLU);
 
-          _backward_kernel = [](const IPortableTensor *output, const IPortableTensor *incoming,
-                                IPortableTensor *outgoing) {
-            nnfw::cker::train::ReLUGrad(getShape(output), getBuffer<float>(output),
-                                        getShape(incoming), getBuffer<float>(incoming),
-                                        getShape(outgoing), getBuffer<float>(outgoing));
+          auto relu_cker = [&alpha]() {
```
Comment (ragmani) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:63]: `alpha` is a local variable. Please use capture by value.
```suggestion
          auto relu_cker = [alpha]() {
```

Comment (zetwhite) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:63]: I thought reference is fine,
Because this lambda function is invoked right now - when it is defined. 
This lambda function is also used in only local!
Is there anything I miss-understood here?

Anyway, alpha is just float type, removing & is no problem with me.

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc (hunk 1 above)
Comment (zetwhite) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:70]:
```suggestion
          }();
```
lambda is invoked here. 
and it returns a function pointer to relu_cker.

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc (hunk 5 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:63]: Ah, I thought `relu_cker is a lambda, but it's not.

**Review by ragmani — APPROVED** [review]

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc (hunk 5 above)
Comment (zetwhite) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:63]: Ah, definitely.. this using is tricky. :sweat_smile:

**Review by jyoungyun — APPROVED** [review]
LGTM

File: runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc [hunk 6]
```diff
@@ -54,22 +55,31 @@ void ElementwiseActivationLayer::configure(const IPortableTensor *input, IPortab
...
+            if (alpha == std::numeric_limits<float>::infinity())
+              return nnfw::cker::train::ReLUGrad;
+            else if (alpha == 6.0f)
```
Comment (YongseopKim) [inline runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: Q. Is it okay to compare float value with float value? I'm not sure....

Comment (zetwhite) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: Could you explain a little more? 

At first glance, I failed to catch your worries.
I thought It is natural to compare one float with another.

Comment (glistening) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: Generally not, but some range of integer is exactly represented in float format.

For float, it has 24-bit significand. 6 is exactly representalbe.

To make sure,

https://en.wikipedia.org/wiki/Single-precision_floating-point_format

> Integers between 0 and 16777216 can be exactly represented

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: Almost it returns true

Comment (zetwhite) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: Aha, I understood @YongseopKim's point.  
There is a floating-point (representataion) error issue.

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: https://codereview.stackexchange.com/questions/275266/c-compare-floats

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: FYI, this is just comment. Do not count me as the number of review.

Comment (glistening) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: > Almost it returns true but rarely could return false. If you already know this and then code it, it would be okay. But if not, we could consider it. 😀

I don't understand. For the case in range (0 and 16777216), it is always safe to compare as I understand. Maybe I don't understand your point yet.

Comment (ragmani) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: As far as I know, `(6.0f == 6.0f)` is always true. They have the same precision since they are `float` literal.
But when one of them is a double or a long double like `(6.0f == 6.0)`, I'm not sure if they have the same precision.

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: > It seems that both exponent value and fraction value of 6.0 and 6.0f are the same.

This is my point.

Comment (ragmani) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: >> It seems that both exponent value and fraction value of 6.0 and 6.0f are the same.
> This is my point.

I updated about it in `https://github.com/Samsung/ONE/pull/12487#discussion_r1456812278`. In short, if comparing value is 6.0, there is no problem.

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: And I already told the term `approximate equality`.

if we have a consensus, because of value 6, it is okay. I'm curious what sth happens if the floating point's precision is too big.
- https://beta.boost.org/doc/libs/1_68_0/libs/math/doc/html/math_toolkit/float_comparison.html

Comment (YongseopKim) [reply runtime/onert/backend/train/ops/ElementwiseActivationLayer.cc:66]: My comment is not pointing out that this is a problem.
